src/gui/log/logmodel.cpp

```diff
@@ -0,0 +1,192 @@
+ * Copyright (C) 2020  Prince Gupta <jagannatharjun11@gmail.com>
+ * Copyright (C) 2019  sledgehammer999 <hammered999@gmail.com>
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public License
+ * as published by the Free Software Foundation; either version 2
+ * of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ *
+ * In addition, as a special exception, the copyright holders give permission to
+ * link this program with the OpenSSL project's "OpenSSL" library (or with
+ * modified versions of it that use the same license as the "OpenSSL" library),
+ * and distribute the linked executables. You must obey the GNU General Public
+ * License in all respects for all of the code used other than "OpenSSL".  If you
+ * modify file(s), you may extend this exception to your version of the file(s),
+ * but you are not obligated to do so. If you do not wish to do so, delete this
+ * exception statement from your version.
+ */
+
+#include "logmodel.h"
+
+#include <QApplication>
+#include <QDateTime>
+#include <QColor>
+#include <QPalette>
+
+#include "base/global.h"
+
+namespace
+{
+    const int MAX_VISIBLE_MESSAGES = 20000;
+}
+
+BaseLogModel::Message::Message(const QString &time, const QString &message, const QColor &foreground, const Log::MsgType type)
+    : m_time(time)
+    , m_message(message)
+    , m_foregroundRole(foreground)
+    , m_userRole(type)
+{
+}
+
+QVariant BaseLogModel::Message::time() const
+{
+    return m_time;
+}
+
+QVariant BaseLogModel::Message::message() const
+{
+    return m_message;
+}
+
+QVariant BaseLogModel::Message::foregroundRole() const
+{
+    return m_foregroundRole;
+}
+
+QVariant BaseLogModel::Message::userRole() const
+{
+    return m_userRole;
+}
+
+BaseLogModel::BaseLogModel(QObject *parent)
+    : QAbstractListModel(parent)
+    , m_messages(MAX_VISIBLE_MESSAGES)
+{
+}
+
+int BaseLogModel::rowCount(const QModelIndex &) const
+{
+    return m_messages.size();
+}
+
+int BaseLogModel::columnCount(const QModelIndex &) const
+{
+    return 1;
+}
+
+QVariant BaseLogModel::data(const QModelIndex &index, const int role) const
+{
+    if (!index.isValid())
+        return {};
+
+    const int messageIndex = index.row();
+    if ((messageIndex < 0) || (messageIndex >= m_messages.size()))
+        return {};
+
+    const Message &message = m_messages[messageIndex];
+    switch (role) {
+    case Qt::ForegroundRole:
+        return message.foregroundRole();
+    case Qt::UserRole:
```

Comment:
Why didn't you make a specific name for this one (and all its stuff)? Why not "MessageTypeRole"?

Reply:
Also the class members have inconsistent names. Some of them have "role" suffix and the rest has no one.

Reply:
I removed suffixes from Message class, this should make the class independent from "role" handling,
I replaced "DataRole" with "MessageTypeRole", which includes different values for type and foreground in addition to message and time. I included the foreground role in the enum as well since QStyledItemDelegate::paint shouldn't paint anything related to the foreground. TypeRole enum constant provides better information about the type of data as compared to the previously used value of Qt::UserRole, one side effect of that was that had to include `logModel.h` in `logFilterModel.cpp`, but that shouldn't be a problem since there is a clear dependency between these two classes.

Reply:
Oops... unfortunately, I didn't pay enough attention to this earlier. As a result, we got an incorrect dependency here.
The model should not depend on any particular view/delegate. It must always return view-independent data at least for mandatory QDisplayRole (so that any view can display its data at least in simplified text format). You can, of course, object that we will never use this model with any other views, but this is an incorrect reason.
There is also no point in using a custom foreground role. "I included the foreground role in the enum as well since QStyledItemDelegate::paint shouldn't paint anything related to the foreground"  is an incorrect judgment. If you don't return Qt::ForegroundRole it just uses default value. If you want to prevent QStyledItemDelegate::paint from painting text you should clear text in passed parameters.

Reply:
>You can, of course, object that we will never use this model with any other views

By the way, we could give the user the option to select a simplified log view (just not using a custom delegate). This might improve performance a bit, right?

Reply:
> If you want to prevent QStyledItemDelegate::paint from painting text you should clear text in passed parameters.

That's the reason I'm not using DisplayRole, even returning for Qt::ForegroundRole will not do anything since DisplayRole is null and the foreground role we return does only make sense with the our message.
I can't return valid value for DisplayRole, since then QStyledItemDelegate will paint it with Foreground, I want `QStyledItemDelegate` to only paint background since painting those myself was a headache anyway. 
If we were to return a full string with DisplayRole(time + message), it will be useless effort to seperate them. 

> The model should not depend on any particular view/delegate.

I don't understand this logic. Probably I'm inexperience, but you yourself said, that "model should prepare data and delegate should do the painting accordingly"
the whole purpose of data is too model is to store and provide data in "convenience form", 
I think you're trying to remove the dependency between model and deleagate, but I don't think it is possible, they are made and build for eachother(XD) and with that logic you will never able to use QML.

Reply:
>By the way, we could give the user the option to select a simplified log view (just not using a custom delegate). This might improve performance a bit, right?


I don't see any performance improvement with that, the only extra work we do is to paint part of a string separately, that doesn't involve any performance heavy calculation or something on that line.

Reply:
>the whole purpose of the model is to store and provide data in "convenience form",
I think you're trying to remove the dependency between model and delegate

Well, it seems we have a misunderstanding of how Qt Model/View is supposed to work (ideally). It's time to make it clear (for me as well). I will try to do this when I find myself near something more convenient to comment on than my smartphone.

Reply:
IMHO, you want to have model return valid values for ForegroundRole and DisplayRole, but this will just create unnecessary work, first, we will need to separate message and time. Second, I want to use `QStyleItemDelegate::paint` but can't find a way to modify or create a new QModelIndex that fit our needs that is one without DisplayRole and ForegroundRole, 
the second one is the biggest blocker and if we had to not use QStyleItemDelegate::paint, that'll just create unnecessary code which may or may not work everywhere. 
For the record, I tried `QStyle::CE_ItemViewItem` it doesn't paint selection rectangle for items under the mouse.  
@Chocobo1 opinions?

Man, I was thinking this PR will complete today😢

Reply:
Guys, relax, so as not to imitate you, I will not insist on further improving the Model/View architecture in this PR. In reality, we don't have a completely perfect implementation anywhere in the code. And it seems that this is not even always possible.